---
title: "Combined TCGA RF Model"
author: "Britney Pheng"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

suppressPackageStartupMessages({
  # BiocManager
  library(GenomicSuperSignature)
  library(curatedTCGAData)
  library(MultiAssayExperiment)
  library(TCGAutils)
  library(ComplexHeatmap)
  
  # CRAN
  library(tidyverse) # includes dplyr, ggplot2, magrittr, tidyr
  library(magick)
  library(wordcloud)
  library(ztable)
  library(metafolio)
  library(randomForest)
  library(caret)
})
```

### Load RAVmodel

```{r message=FALSE, warning=FALSE}
RAVmodel <- getModel('C2', load=TRUE)
```

```{r}
# Randomly select 70% of samples for training
set.seed(1)
num_sample <- ncol(combined_data)
train_sample_ind <- sample(seq_len(num_sample), round(num_sample*0.7))

meta_train <- combined_data_meta[train_sample_ind,] 
data_train <- combined_data[, train_sample_ind] # 20,501 genes x 612 samples x 3205 attributes
```

Remove batch effect variables from metadata table

```{r}
batch_var <- "analyte|portion|procurement|aliquot|uuid|barcode"
batch_ind <- grep(batch_var, colnames(meta_train))
meta_train <- meta_train[,-batch_ind] # 612 samples x 1064 metadata attributes
```

```{r}
meta_test <- combined_data_meta[-train_sample_ind,-batch_ind]
data_test <- combined_data[, -train_sample_ind]
```

## Based on variable types

```{r}
## Check for data types in listData
var_type <- meta_train@listData
unique(sapply(var_type, type))

## Separate training data's metadata into two subsets: 
## character variables (~ categorical) and numeric variables (~ continuous)
charcTb <- meta_train[, sapply(var_type, class) == 'character']
numTb <- meta_train[, sapply(var_type, class) %in% c('numeric', 'integer')]
```

## Based on variable types; Test Data

```{r}
## Check for data types in listData
var_type_testdata <- meta_test@listData

## Separate test data's metadata into two subsets: 
## character variables (~ categorical) and numeric variables (~ continuous)
charcTb_testdata <- meta_test[, sapply(var_type_testdata, class) == 'character']
numTb_testdata <- meta_test[, sapply(var_type_testdata, class) %in% c('numeric', 'integer')]
```

# Sample Score
```{r}
## Calculate validation scores
sampleScore <- calculateScore(data_train, RAVmodel)
rownames(sampleScore) <- gsub("\\.", "-", rownames(sampleScore))

## Test data: calculate validation scores
sampleScore_testdata <- calculateScore(data_test, RAVmodel)
rownames(sampleScore_testdata) <- gsub("\\.", "-", rownames(sampleScore_testdata))

```

Sample scores only for the validated RAVs

```{r}
## Training Data
validate_data <- validate(data_train, RAVmodel)
heatmapTable(validate_data, RAVmodel)

validated_ind <- validatedSignatures(validate_data, 
                                     RAVmodel, 
                                     num.out = 4764, 
                                     #scoreCutoff = 0.45, 
                                     indexOnly = TRUE)

## Subset sampleScore
sampleScore_sub <- sampleScore[,validated_ind] %>% as.data.frame()

## Test Data
validate_testdata <- validate(data_test, RAVmodel)
heatmapTable(validate_testdata, RAVmodel, num.out = 15)

validated_ind_testdata <- validatedSignatures(validate_testdata, 
                                     RAVmodel, 
                                     num.out = 4764, 
                                     #scoreCutoff = 0.45, 
                                     indexOnly = TRUE)

## Subset sampleScore
sampleScore_sub_testdata <- sampleScore_testdata[,validated_ind_testdata] %>% as.data.frame()
```

```{r}
## Convert character variables into the factor data type
factorTb <- charcTb
factorTb[sapply(factorTb, is.character)] <- lapply(factorTb[sapply(factorTb, is.character)], factor)

levels(factorTb@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status)[levels(factorTb@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status) == "msi-h"] <- "msi"
```

```{r}
## Convert character variables into the factor data type
factorTb_testdata <- charcTb_testdata
factorTb_testdata[sapply(factorTb_testdata, is.character)] <- lapply(factorTb_testdata[sapply(factorTb_testdata, is.character)], factor)

levels(factorTb_testdata@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status)[levels(factorTb_testdata@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status) == "msi-h"] <- "msi"
```

# Prediction models

## Use top 15 validated RAVs for TCGA-COAD,STAD, UCEC

```{r}
# validated_RAVs <- validated_ind %>% as.vector() %>% as.character()
# validated_RAVs <- paste("RAV", validated_RAVs, sep="")
validated_RAVs <- c("RAV357", "RAV27", "RAV834", "RAV190", "RAV1166", "RAV517",
                    "RAV2344", "RAV324", "RAV438", "RAV220", "RAV868", "RAV1008", "RAV625")
target_attr <- "patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status"

labels <- factorTb[[target_attr]]
nonNALabels <- which(!is.na(labels))
data <- sampleScore_sub[,validated_RAVs]

train_data <- data[nonNALabels,]
train_labels <- labels[nonNALabels]
train_data$MSI_Status <- train_labels
```

## Random Forest classification model

```{r}
rf <- randomForest(MSI_Status~., data=train_data, proximity=TRUE)
print(rf)

p1 <- predict(rf, train_data)
#confusionMatrix(p1, train_data$MSI_Status)

confusion_matrix <- table(p1, train_data$MSI_Status)
confusion_matrix

accuracy <- sum(diag(confusion_matrix)) / sum(confusion_matrix)
print(accuracy)

importance(rf)
varImpPlot(rf)
```

```{r}
predictions1 <- predict(rf, newdata = train_data, predict.all = TRUE)

tree_predictions <- predictions1$individual  # Extract tree predictions

# Calculate accuracy for each tree (for classification)
tree_accuracy <- apply(tree_predictions, 2, function(tree_pred) {
  mean(tree_pred == train_data$MSI_Status)
})

# print(tree_accuracy)
# sort(tree_accuracy)

# Calculate weights proportional to accuracy
weights <- tree_accuracy / sum(tree_accuracy)

# Combine predictions using weighted average
weighted_predictions <- apply(tree_predictions, 1, function(preds) {
  weighted_pred <- tapply(weights, preds, sum)
  names(which.max(weighted_pred))
})

# Convert weighted predictions to factor
weighted_predictions <- as.factor(weighted_predictions)

confusion_matrix <- table(weighted_predictions, train_data$MSI_Status)
confusion_matrix
accuracy <- sum(diag(confusion_matrix)) / sum(confusion_matrix)
print(accuracy)


```

## Test Data

```{r}
labels <- factorTb_testdata[[target_attr]]
nonNALabels <- which(!is.na(labels))
data <- sampleScore_sub_testdata[,validated_RAVs]

test_data <- data[nonNALabels,]
test_labels <- labels[nonNALabels]

test_data$MSI_Status <- test_labels
```

```{r}
p2 <- predict(rf, test_data)
confusionMatrix(p2, test_data$MSI_Status)
```

```{r}
library("rpart")
library("rpart.plot")

#fit new model with rpart
train_msi_model <- rpart(MSI_Status ~., data = train_data)
test_msi_model <- rpart(MSI_Status ~., data = test_data)

# plot the tree
rpart.plot(train_msi_model)
```

```{r}
plot(rf)
```

