---
title: "MSI Random Forest Model - curatedCRCData"
author: "Britney Pheng"
date: "`r format(Sys.time(), '%B %d, %Y')`"
format:
    html:
        fontsize: 14pxs
        toc: true
        top-depth: 3
output: html_document
editor: visual
---

Quarto enables you to weave together content and executable code into a finished document. To learn more about Quarto see <https://quarto.org>.

# Initial Setup

## Load packages

```{r packages, warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  # BiocManager
  library(GenomicSuperSignature)
  library(curatedTCGAData)
  library(MultiAssayExperiment)
  library(TCGAutils)
  library(ComplexHeatmap)
  library(SummarizedExperiment)
  library(curatedCRCData)
  
  # CRAN
  library(tidyverse) # includes dplyr, ggplot2, magrittr, tidyr
  library(magick)
  library(wordcloud)
  library(ztable)
  library(metafolio)
  library(randomForest)
  library(caret)
})
```

## Load data

```{r}
train_rf_data <- read.csv("/Users/bpheng/data/meta_train_char_cancer.csv")
sampleScore_sub <- read.csv("/Users/bpheng/data/sampleScore_train_cancer.csv")
```

```{r}
## Convert character variables into the factor data type
factorTb <- train_rf_data
factorTb[sapply(factorTb, is.character)] <- lapply(factorTb[sapply(factorTb, is.character)], factor)

levels(factorTb$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status)[levels(factorTb$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status) == "msi-h"] <- "msi"

#levels(factorTb$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status)
```

# Prediction models

## Use top 15 validated RAVs for TCGA-COAD

```{r}
validated_RAVs <- c("RAV61", "RAV64", "RAV188", "RAV190", "RAV192", 
                    "RAV220", "RAV324", "RAV438", "RAV833", "RAV834", 
                    "RAV868", "RAV1008", "RAV1575")
target_attr <- "patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status"

labels <- factorTb[[target_attr]]
nonNALabels <- which(!is.na(labels))
data <- sampleScore_sub[,validated_RAVs]

train_data <- data[nonNALabels,]
train_labels <- labels[nonNALabels]
train_data$MSI_Status <- train_labels
```

## Random Forest classification model

```{r}
rf <- randomForest(MSI_Status~., data=train_data, proximity=TRUE)
#print(rf)
```

#UCEC

```{r}
ucec <- curatedTCGAData(diseaseCode = 'UCEC',
                        assays = 'RNA*',
                        version = '2.0.1',
                        dry.run = FALSE)

## Parse out cancer vs normal samples
sampleTables(ucec)
ucec2 <- TCGAsplitAssays(ucec, c("01", "11"))

ucec_rna_cancer <- getWithColData(ucec2,
                           '01_UCEC_RNASeq2Gene-20160128')
assay(ucec_rna_cancer) <- log2(assay(ucec_rna_cancer) + 1)

ucec_rna_cancer <- ucec_rna_cancer[,ucec_rna_cancer@colData@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status %in% c('msi-h', 'mss')]

ucec_meta_cancer <- colData(ucec_rna_cancer)
```

```{r}
ucec.meta <- colData(ucec_rna_cancer)
ucec.sampleScore <- calculateScore(ucec_rna_cancer, RAVmodel) %>% as.data.frame()
```

```{r}
target_attr <- "patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status"

var_type <- ucec.meta@listData
unique(sapply(var_type, type))
charcTb <- ucec.meta[, sapply(var_type, class) == 'character']
factorTb <- charcTb
factorTb[sapply(factorTb, is.character)] <- lapply(factorTb[sapply(factorTb, is.character)], factor)

levels(factorTb@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status)[levels(factorTb@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status) == "msi-h"] <- "msi"

labels <- factorTb[[target_attr]]
nonNALabels <- which(!is.na(labels))
ucec_data <- ucec.sampleScore[,validated_RAVs]

new_ucec_data <- ucec_data[nonNALabels,]
new_labels <- labels[nonNALabels]
new_ucec_data$msi <- new_labels

```

```{r}
ucec.prediction <- predict(rf, new_ucec_data)
confusionMatrix(ucec.prediction, new_ucec_data$msi)
```

```{r}
library("rpart")
library("rpart.plot")

#fit new model with rpart
msi_ucec <- rpart(msi ~., data = new_ucec_data)

# plot the tree
rpart.plot(msi_ucec)
```
