---
title: "MSI Random Forest Model - STAD"
author: "Britney Pheng"
date: "`r format(Sys.time(), '%B %d, %Y')`"
format:
    html:
        fontsize: 14pxs
        toc: true
        top-depth: 3
output: html_document
editor: visual
---

# Initial Setup

## Load packages

```{r packages, warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  # BiocManager
  library(GenomicSuperSignature)
  library(curatedTCGAData)
  library(MultiAssayExperiment)
  library(TCGAutils)
  library(ComplexHeatmap)
  library(SummarizedExperiment)
  library(curatedCRCData)
  
  # CRAN
  library(tidyverse) # includes dplyr, ggplot2, magrittr, tidyr
  library(magick)
  library(wordcloud)
  library(ztable)
  library(metafolio)
  library(randomForest)
  library(caret)
})
```

## Load RAVmodel

```{r RAV, message=FALSE, warning=FALSE}
RAVmodel <- getModel('C2', load=TRUE)
```

## Load data

```{r}
train_rf_data <- read.csv("/Users/bpheng/data/meta_train_char_cancer.csv")
sampleScore_sub <- read.csv("/Users/bpheng/data/sampleScore_train_cancer.csv")
```

## Load RAVmodel

```{r message=FALSE, warning=FALSE}
RAVmodel <- getModel('C2', load=TRUE)
```

```{r}
## Convert character variables into the factor data type
factorTb <- train_rf_data
factorTb[sapply(factorTb, is.character)] <- lapply(factorTb[sapply(factorTb, is.character)], factor)

levels(factorTb$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status)[levels(factorTb$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status) == "msi-h"] <- "msi"

#levels(factorTb$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status)
```

# Prediction models

## Use top 15 validated RAVs for TCGA-COAD

```{r}
validated_RAVs <- c("RAV61", "RAV64", "RAV188", "RAV190", "RAV192", 
                    "RAV220", "RAV324", "RAV438", "RAV833", "RAV834", 
                    "RAV868", "RAV1008", "RAV1575")
target_attr <- "patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status"

labels <- factorTb[[target_attr]]
nonNALabels <- which(!is.na(labels))
data <- sampleScore_sub[,validated_RAVs]

train_data <- data[nonNALabels,]
train_labels <- labels[nonNALabels]
train_data$MSI_Status <- train_labels
```

## Random Forest classification model

```{r}
rf <- randomForest(MSI_Status~., data=train_data, proximity=TRUE)
#print(rf)
```

#STAD

```{r}
stad <- curatedTCGAData(diseaseCode = 'STAD',
                        assays = 'RNA*',
                        version = '2.0.1',
                        dry.run = FALSE)

## Parse out cancer vs normal samples
sampleTables(stad)
stad2 <- TCGAsplitAssays(stad, c("01", "11"))

stad_rna_cancer <- getWithColData(stad2,
                           '01_STAD_RNASeq2Gene-20160128')
assay(stad_rna_cancer) <- log2(assay(stad_rna_cancer) + 1)

stad_rna_cancer <- stad_rna_cancer[,stad_rna_cancer@colData@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status %in% c('msi-h', 'mss')]

stad_meta_cancer <- colData(stad_rna_cancer)
```

```{r}
stad_meta <- colData(stad_rna_cancer)
stad_sampleScore <- calculateScore(stad_rna_cancer, RAVmodel) %>% as.data.frame()
```

```{r}
target_attr <- "patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status"

var_type <- stad_meta@listData
unique(sapply(var_type, type))
charcTb <- stad_meta[, sapply(var_type, class) == 'character']
factorTb <- charcTb
factorTb[sapply(factorTb, is.character)] <- lapply(factorTb[sapply(factorTb, is.character)], factor)

levels(factorTb@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status)[levels(factorTb@listData$patient.microsatellite_instability_test_results.microsatellite_instability_test_result.mononucleotide_and_dinucleotide_marker_panel_analysis_status) == "msi-h"] <- "msi"

labels <- factorTb[[target_attr]]
nonNALabels <- which(!is.na(labels))
data <- stad_sampleScore[,validated_RAVs]

new_data <- data[nonNALabels,]
new_labels <- labels[nonNALabels]
new_data$msi <- new_labels

```

```{r}
stad_prediction <- predict(rf, new_data)
stad_matrix <- confusionMatrix(stad_prediction, new_data$msi)
stad_df <- stad_matrix$table

mosaic(stad_df,legend=FALSE,
       main = "STAD Confusion Matrix Mosaic Plot")
```

```{r}
library("rpart")
library("rpart.plot")

#fit new model with rpart
msi_stad <- rpart(msi ~., data = new_data)

# plot the tree
rpart.plot(msi_stad)
```
